File: Recursion/permutations_1.cpp
/*
https://leetcode.com/problems/permutations/description/

46. Permutations

Given an array nums of distinct integers, return all the possible
permutations. You can return the answer in any order.

Example 1:

Input: nums = [1,2,3]
Output: [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]

Example 2:

Input: nums = [0,1]
Output: [[0,1],[1,0]]

Example 3:

Input: nums = [1]
Output: [[1]]

Constraints:

1 <= nums.length <= 6
-10 <= nums[i] <= 10
All the integers of nums are unique.
*/

#include <iostream>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution
{
private:
  void dfs(const vector<int>& nums, vector<int>& temp, vector<bool>& visited,
           vector<vector<int>>& result)
  {
    if (temp.size() == nums.size())
    {
      result.push_back(temp);
      return;
    }
    for (int i = 0; i < nums.size(); i++)
    {
      if (!visited[i])
      {
        visited[i] = true;
        temp.push_back(nums[i]);
        dfs(nums, temp, visited, result);
        temp.pop_back();
        visited[i] = false;
      }
    }
  }

  void dfs(vector<int>& nums, int index, vector<vector<int>>& result)
  {
    if (index == nums.size())
    {
      result.push_back(nums);
      return;
    }
    for (int i = index; i < nums.size(); i++)
    {
      swap(nums[index], nums[i]);
      dfs(nums, index + 1, result);
      swap(nums[index], nums[i]);
    }
  }

public:
  vector<vector<int>> permute1(vector<int>& nums)
  {
    vector<vector<int>> result;
    vector<bool> visited(nums.size(), false);
    vector<int> temp;
    dfs(nums, temp, visited, result);
    return result;
  }

  vector<vector<int>> permute2(vector<int>& nums)
  {
    vector<vector<int>> result;
    dfs(nums, 0, result);
    return result;
  }
};

int main()
{
  cout << "This program generates all possible permutations of a given array "
          "of distinct integers.\n";
  cout << "Enter the size of the array (1 <= size <= 6): ";

  int n;
  cin >> n;

  vector<int> nums(n);
  cout << "Enter " << n << " distinct integers (-10 <= value <= 10): ";
  for (int i = 0; i < n; ++i)
  {
    cin >> nums[i];
  }

  Solution sol;
  vector<vector<int>> result = sol.permute1(nums);
  cout << "\nPermutations using visited tracking:\n";
  for (const auto& perm : result)
  {
    for (int num : perm)
    {
      cout << num << " ";
    }
    cout << endl;
  }

  vector<vector<int>> result2 = sol.permute2(nums);
  cout << "\nPermutations using in-place swapping:\n";
  for (const auto& perm : result2)
  {
    for (int num : perm)
    {
      cout << num << " ";
    }
    cout << endl;
  }

  return 0;
}

File: recursion/Permutations1.java
/*
 * https://leetcode.com/problems/permutations/description/
 * https://youtu.be/YK78FU5Ffjw
 * https://youtu.be/f2ic2Rsc9pU
 */

package recursion;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Permutations1 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int[] nums = new int[n];
        for (int i = 0; i < n; i++) {
            nums[i] = sc.nextInt();
        }

        Permutations1 sol = new Permutations1();
        System.out.println(sol.permute2(nums));

        sc.close();
    }

    /**
     * Permutation using swapping on each index
     */
    public List<List<Integer>> permute2(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> curr = new ArrayList<>();

        int n = nums.length;

        backtrack(nums, n, 0, curr, res);

        return res;
    }

    private void backtrack(int[] nums, int n, int index, List<Integer> curr, List<List<Integer>> res) {
        if (index == n) {
            res.add(new ArrayList<>(curr));
            return;
        }

        for (int i = index; i < n; i++) {
            swap(nums, index, i);
            curr.add(nums[index]); // We are playing on current index only, swapping it with others
            backtrack(nums, n, index + 1, curr, res); // Understand TODO: In subsequence questions we did i + 1, here we
                                                      // are doing index + 1, we are dealing with each index at a time
            curr.remove(curr.size() - 1);
            swap(nums, index, i);
        }
    }

    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }

    /**
     * Permutation using visited hash
     */
    public List<List<Integer>> permute1(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> curr = new ArrayList<>();

        int n = nums.length;
        boolean[] visited = new boolean[n];

        backtrack(nums, n, visited, curr, res);

        return res;
    }

    private void backtrack(int[] nums, int n, boolean[] visited, List<Integer> curr,
            List<List<Integer>> res) {

        if (curr.size() == n) {
            res.add(new ArrayList<>(curr));
            return;
        }

        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                visited[i] = true;
                curr.add(nums[i]);
                backtrack(nums, n, visited, curr, res);
                curr.remove(curr.size() - 1);
                visited[i] = false;
            }
        }
    }
}

// Recursive Time Complexity: O(n!) At first level i have N choices, the N -
// 1
// choices, .... then 1 choice

// Recursive Space Complexity: O(N)

// Overall Time Complexity: O(N! * N): Each permutaion is copied to result
